### PR TITLE
Fix endless loop bug

### LINKED ISSUE #543: RAM exhaustion on reading binary GOTO
the attached (likely incorrect) binary GOTO of 48 bytes causes CBMC to consume > 8G of RAM on reading

[endless-loop-8G.zip](https://github.com/diffblue/cbmc/files/772197/endless-loop-8G.zip)

### PR DESCRIPTION
**Pull request by reuk**
Fixes #543

### CODE REVIEW

File: src/util/irep_serialization.cpp [hunk 1]
```diff
@@ -327,7 +327,7 @@ irep_idt irep_serializationt::read_gb_string(std::istream &in)
   char c;
   size_t length=0;
 
-  while((c = in.get()) != 0)
+  while(in.good() && (c = in.get()) != 0)
```
Comment (tautschnig) [inline src/util/irep_serialization.cpp]: Shouldn't an error be reported if such a string without 0-termination is encountered?

**Comment by tautschnig** [comment]
> error reporting would be nice, but this is from fuzzed binary GOTO, so it is very unlikely to appear in the wild

This patch would move the status from denial-of-service (not great, but low risk of exposing private data) to possibly arbitrary crashes at a later stage that may include buffer overflows. Error detection on invalid input MUST happen as early as possible.

**Comment by mgudemann** [comment]
@tautschnig I see your point, I agree, the error should be detected here and execution should not be continued in this situation.

File: src/util/irep_serialization.cpp [hunk 2]
```diff
@@ -324,11 +324,15 @@ Function: irep_serializationt::read_gb_string
...
 {
-  char c;
   size_t length=0;
 
-  while((c = in.get()) != 0)
+  while(char c = in.get())
   {
+    if(!in.good())
+    {
+      throw irep_serialization_errort("gb string is not null-terminated");
```
Comment (tautschnig) [inline src/util/irep_serialization.cpp]: It seems Visual Studio doesn't like this. Please just throw the C string, no need for `irep_serialization_errort` at this point.

Comment (reuk) [inline src/util/irep_serialization.cpp]: It's just that VS doesn't understand C++11. Better to fix the exception constructor than to throw a char star though surely? 
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#e14-use-purpose-designed-user-defined-types-as-exceptions-not-built-in-types

Comment (tautschnig) [inline src/util/irep_serialization.cpp]: Not until #911 is merged, or you change all other 1583 uses of `throw` without a user-defined type.

File: src/util/irep_serialization.cpp [hunk 3]
```diff
@@ -324,11 +328,15 @@ Function: irep_serializationt::read_gb_string
 
 irep_idt irep_serializationt::read_gb_string(std::istream &in)
 {
-  char c;
   size_t length=0;
 
-  while((c = in.get()) != 0)
+  while(char c = in.get())
```
Comment (peterschrammel) [inline src/util/irep_serialization.cpp]: spacing around =

File: src/util/irep_serialization.cpp [hunk 4]
```diff
@@ -14,6 +14,10 @@ Date: May 2007
 #include "irep_serialization.h"
 #include "string_hash.h"
 
+irep_serialization_errort::irep_serialization_errort(
+  const std::string &error)
+    : std::runtime_error(error) {}
```
Comment (peterschrammel) [inline src/util/irep_serialization.cpp]: coding style for `:`

File: src/util/irep_serialization.cpp [hunk 5]
```diff
@@ -225,11 +230,16 @@ void write_gb_string(std::ostream &out, const std::string &s)
...
 {
-  char c;
+  char c='\0';
   size_t length=0;
 
   while((c=static_cast<char>(in.get()))!=0)
   {
+    if(!in.good())
+    {
+      throw irep_serialization_errort("gb string is not null-terminated");
```
Comment (peterschrammel) [inline src/util/irep_serialization.cpp]: 'gb' is not very user friendly.

**Comment by tautschnig** [comment]
@peterschrammel Could you please review whether you are happy with the current state?

**Comment by tautschnig** [comment]
@peterschrammel ping?

**Pull request closed without merging**